### PR TITLE
Update shared HPackEncoder

### PR DESCRIPTION
**Pull request by JamesNK**
Sync changes from https://github.com/dotnet/aspnetcore/pull/19393. Review now, but wait until that PR is merged before merging this.

Optimizing the enumerator eliminates allocations per HTTP request.

### CODE REVIEW

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs [hunk 1]
```diff
@@ -2,26 +2,30 @@
...
 
 namespace System.Net.Http.HPack
 {
     internal class HPackEncoder
     {
-        private IEnumerator<KeyValuePair<string, string>> _enumerator;
+        private HeadersEnumerator _enumerator;
 
-        public bool BeginEncode(IEnumerable<KeyValuePair<string, string>> headers, Span<byte> buffer, out int length)
+        public bool BeginEncode(HeadersEnumerator enumerator, Span<byte> buffer, out int length)
```
Comment (Tratcher) [inline src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: I thought the point was that HttpClient didn't need these APIs and they could be removed?

Comment (JamesNK) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: HttpClient might not use this code but non-Kestrel code in our repo does, e.g. HttpSys tests, httpcat.

I changed my PR to use an `#if`. It minimizes the amount of overall change.

Comment (Tratcher) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: Minimizing changes is not critical at this point, I'd rather have a clean result. If only test code needs these old APIs then is there something straightforward we can do to restructure how they call this?

In the design for dynamic header compression we've already proposed changes that would require moving this enumerator outside of HPackEncoder. Maybe now is a good time to separate them.

Comment (JamesNK) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: > If only test code needs these old APIs

Kestrel **and** tests use these APIs. My previous copied and pasted class that wrapped HPackEncoder still was the same API. I have just changed separation of our custom enumerator to use an `#if`.

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs [hunk 2]
```diff
@@ -2,26 +2,30 @@
 // Licensed under the Apache License, Version 2.0.
 // See THIRD-PARTY-NOTICES.TXT in the project root for license information.
 
-using System.Collections.Generic;
 using System.Diagnostics;
+#if KESTREL
+using HeadersEnumerator = Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2HeadersEnumerator;
+#else
+using HeadersEnumerator = System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, string>>;
```
Comment (jkotalik) [inline src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: Do you have a corresponding PR to runtime to make sure it compiles there?

Comment (jkotalik) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: nvrm I'm blind...

Comment (JamesNK) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: This is the PR to runtime?

Comment (jkotalik) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs]: yeah I realize that now... oops.

**Review by Tratcher — COMMENTED** [review]
Approved with the understanding that you plan to rewrite/remove these APIs when working on static compression.

**Review by Tratcher — APPROVED** [review]

**Comment by JamesNK** [comment]
I don't have write permissions. Someone who does will need to merge it.

**Review by scalablecory — APPROVED** [review]
It looks fine, but I think it'd be worth looking at why the encoder needs to hold a copy of the enumerator -- from high-level view it seems like this API can go away, the field can go away, and the class made static (until we add dynamic table encoding, then it'll need some other state).

**Comment by Tratcher** [comment]
> It looks fine, but I think it'd be worth looking at why the encoder needs to hold a copy of the enumerator -- from high-level view it seems like this API can go away

Agreed. That's being investigated in the next PR. https://github.com/dotnet/aspnetcore/pull/19521

**Pull request merged**